qiskit_machine_learning/kernels/quantum_kernel.py

```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+            adjust_feature_map: If ``True`` tried to adjust the number of qubits of the feature map.
 
+        Raises:
+            ValueError: If the size of the vector is not the same as the number of parameters of the
+                feature map or when the feature map does not allow to change the number of qubits.
+        """
+        if vec is not None and len(vec) != self._feature_map.num_parameters:
+            raise_error = True
+            if adjust_feature_map:
+                try:
+                    # an attempt to update the feature map with the required number of qubits
+                    self._feature_map.num_qubits = vec
+                    raise_error = False
+                except AttributeError:
+                    pass
+
+            if raise_error:
+                raise ValueError(
+                    f"{vec_name} and class feature map have incompatible dimensions.\n"
+                    f"{vec_name} has {len(vec)} dimensions, "
+                    f"but feature map has {self._feature_map.num_parameters}."
+                )
+
+    def _calculate_kernel_statevector(
+        self, x_vec: np.ndarray, y_vec: np.ndarray, is_symmetric: bool, kernel: np.ndarray
+    ):
         # get indices to calculate
+        row_indices, col_indices = self._get_indices(x_vec, y_vec, is_symmetric)
+
         if is_symmetric:
-            mus, nus = np.triu_indices(x_vec.shape[0], k=1)  # remove diagonal
-        else:
-            mus, nus = np.indices((x_vec.shape[0], y_vec.shape[0]))
-            mus = np.asarray(mus.flat)
-            nus = np.asarray(nus.flat)
+            to_be_computed_data = x_vec
+        else:  # not symmetric
+            to_be_computed_data = np.concatenate((x_vec, y_vec))
+
+        feature_map_params = ParameterVector("par_x", self._feature_map.num_parameters)
+        parameterized_circuit = self._construct_circuit_statevector(
+            feature_map_params,
+            feature_map_params,
+        )
+        parameterized_circuit = self._quantum_instance.transpile(
+            parameterized_circuit, pass_manager=self._quantum_instance.unbound_pass_manager
+        )[0]
+        statevectors = []
+
+        for min_idx in range(0, len(to_be_computed_data), self._batch_size):
+            max_idx = min(min_idx + self._batch_size, len(to_be_computed_data))
+            circuits = [
+                parameterized_circuit.assign_parameters({feature_map_params: x})
+                for x in to_be_computed_data[min_idx:max_idx]
+            ]
+            if self._quantum_instance.bound_pass_manager is not None:
+                circuits = self._quantum_instance.transpile(
+                    circuits, pass_manager=self._quantum_instance.bound_pass_manager
+                )
+            results = self._quantum_instance.execute(circuits, had_transpiled=True)
+            for j in range(max_idx - min_idx):
+                statevectors.append(results.get_statevector(j))
```

Comment:
I would extract it to a method.

Reply:
I'd prefer to keep as it's now for consistency with `_calculate_kernel()`.